### PR TITLE
feat: add babel plugin for web builder

### PR DESCRIPTION
**Pull request by sanyuan0704**
# PR Details

## Description
1. Add `modifyPresetEnvOptions` and `modifyPresetReactOptions` in @modern-js/babel-preset-base
2. Repair upstream and downstream relationships about @modern-js/babel-preset-base and @modern-js/core
3. Add babel plugin for web builder.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/web-builder/package.json [hunk 1]
```diff
@@ -36,17 +36,24 @@
     "test:watch": "NODE_OPTIONS=--experimental-vm-modules jest --watch"
   },
   "dependencies": {
-    "@modern-js/utils": "workspace:^1.7.12",
     "ajv": "^8.11.0",
-    "css-minimizer-webpack-plugin": "4.0.0",
     "lodash": "^4.17.21",
+    "@modern-js/babel-preset-app": "workspace:*",
+    "@modern-js/utils": "workspace:*",
```
Comment (chenjiahan) [inline packages/builder/web-builder/package.json:42]: 现在 modern.js 内部的包是不 lock 版本的哈

一般 dependencies 会写成 "workspace:^x.y.z" （具体的版本号）

devDependencies 可以写成 "workspace:*"

File: packages/builder/web-builder/src/core/createContext.ts [hunk 2]
```diff
@@ -74,6 +76,7 @@ export function createPublicContext(
       'configPath',
       'tsconfigPath',
       'originalConfig',
+      'metaName',
```
Comment (chenjiahan) [inline packages/builder/web-builder/src/core/createContext.ts:79]: 可以考虑下有没有比 metaName 更准确的命名 😂

File: packages/cli/babel-preset-app/package.json [hunk 3]
```diff
@@ -39,8 +39,8 @@
     "@babel/plugin-transform-destructuring": "^7.18.0",
     "@babel/runtime": "^7.18.0",
     "@babel/types": "^7.18.0",
-    "@modern-js/babel-preset-base": "workspace:^1.5.0",
-    "@modern-js/utils": "workspace:^1.8.1",
+    "@modern-js/babel-preset-base": "workspace:*",
```
Comment (chenjiahan) [inline packages/cli/babel-preset-app/package.json:42]: 同上

File: packages/cli/babel-preset-base/package.json [hunk 4]
```diff
@@ -43,20 +43,19 @@
     "@babel/preset-react": "^7.17.12",
     "@babel/preset-typescript": "^7.17.12",
     "@babel/runtime": "^7.18.0",
+    "@modern-js/utils": "workspace:*",
```
Comment (chenjiahan) [inline packages/cli/babel-preset-base/package.json:46]: 同上

File: packages/cli/babel-preset-base/src/babel-utils.ts [hunk 5]
```diff
@@ -75,6 +75,22 @@ const removePresets = (
   });
 };
 
+const modifyPresetOptions = <T>(
+  presetName: string,
+  options: T,
+  presets: PluginItem[] = [],
+) => {
+  presets.forEach((preset: any) => {
+    // Assume preset has [name, options] structure
```
Comment (chenjiahan) [inline packages/cli/babel-preset-base/src/babel-utils.ts:84]: 这里可以做一下容错，用户新增的 preset 可能不是这个结构

File: packages/cli/babel-preset-base/src/babel-utils.ts [hunk 6]
```diff
@@ -75,6 +75,22 @@ const removePresets = (
   });
 };
 
+const modifyPresetOptions = <T>(
+  presetName: string,
+  options: T,
+  presets: PluginItem[] = [],
+) => {
+  presets.forEach((preset: any) => {
```
Comment (chenjiahan) [inline packages/cli/babel-preset-base/src/babel-utils.ts:83]: any 最好干掉

File: packages/builder/web-builder/src/types/config/output.ts [hunk 7]
```diff
@@ -19,4 +19,14 @@ export interface OutputConfig {
   disableMinimize?: boolean;
   disableSourceMap?: boolean;
   disableFilenameHash?: boolean;
+  enableLatestDecorators?: boolean;
+  polyfill?: Polyfill;
+}
+
+export type Polyfill = 'usage' | 'entry' | 'ua' | 'off';
+
+export interface WebBuilderOutputConfig {
```
Comment (chenjiahan) [inline packages/builder/web-builder/src/types/config/output.ts:28]: 这里应该是合并代码的时候重复了

File: packages/builder/web-builder/src/core/createContext.ts (hunk 2 above)
Comment (sanyuan0704) [reply packages/builder/web-builder/src/core/createContext.ts:79]: framework 怎么样

Comment (chenjiahan) [reply packages/builder/web-builder/src/core/createContext.ts:79]: 可以滴

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**